File: app/scraper/parsers.py
import re
from lxml import html
from typing import Optional


def extract_title(tree: html.HtmlElement) -> Optional[str]:
    el = tree.xpath('//*[@id="productTitle"]/text()')
    if el:
        return el[0].strip()
    el = tree.xpath('//span[@id="productTitle"]/text()')
    if el:
        return el[0].strip()
    return None


def extract_price(tree: html.HtmlElement) -> Optional[str]:
    selectors = [
        '//span[contains(@class,"priceToPay")]//span[contains(@class,"a-price-whole")]/text()',
        '//*[@id="corePrice_feature_div"]//span[contains(@class,"a-price-whole")]/text()',
        '//span[@id="priceblock_ourprice"]/text()',
        '//span[@id="priceblock_dealprice"]/text()',
        '//*[@id="apex_offerDisplay_desktop"]//span[contains(@class,"a-price-whole")]/text()',
        '//span[contains(@class,"a-price")]//span[contains(@class,"a-price-whole")]/text()',
    ]
    for sel in selectors:
        result = tree.xpath(sel)
        if result:
            price_str = result[0].strip().replace(",", "")
            if price_str:
                fraction = tree.xpath(sel.replace("a-price-whole", "a-price-fraction"))
                if fraction:
                    return f"₹{price_str}.{fraction[0].strip()}"
                return f"₹{price_str}"
    return None


def extract_rating(tree: html.HtmlElement) -> Optional[str]:
    el = tree.xpath('//*[@id="acrPopover"]/@title')
    if el:
        match = re.search(r"([\d.]+)", el[0])
        if match:
            return match.group(1)
    el = tree.xpath('//span[@data-hook="rating-out-of-text"]/text()')
    if el:
        match = re.search(r"([\d.]+)", el[0])
        if match:
            return match.group(1)
    el = tree.xpath('//i[contains(@class,"a-icon-star")]//span/text()')
    if el:
        match = re.search(r"([\d.]+)", el[0])
        if match:
            return match.group(1)
    return None


def extract_rating_count(tree: html.HtmlElement) -> Optional[str]:
    selectors = [
        '//*[@id="acrCustomerReviewText"]/text()',
        '//*[@id="acrCustomerReviewLink"]//span/text()',
        '//span[@data-hook="total-review-count"]/text()',
        '//*[@id="acrPopover"]/..//span[contains(@class,"a-size-base")]/text()',
    ]
    for sel in selectors:
        els = tree.xpath(sel)
        for el in els:
            match = re.search(r"([\d,]+)", el)
            if match:
                return match.group(1).replace(",", "")
    return None


def extract_bsr(tree: html.HtmlElement) -> Optional[str]:
    sales_rank = tree.xpath('//*[@id="SalesRank"]//text()')
    if sales_rank:
        text = " ".join(t.strip() for t in sales_rank if t.strip())
        match = re.search(r"#([\d,]+)\s+in\s+(.+?)(?:\(|$)", text)
        if match:
            rank = match.group(1).replace(",", "")
            category = match.group(2).strip()
            return f"#{rank} in {category}"

    tables = tree.xpath(
        '//table[contains(@id,"productDetails")]//tr | '
        '//div[@id="detailBulletsWrapper_feature_div"]//li'
    )
    for row in tables:
        text = " ".join(row.xpath('.//text()')).strip()
        if "Best Sellers Rank" in text or "best seller" in text.lower():
            match = re.search(r"#([\d,]+)\s+in\s+(.+?)(?:\(|$)", text)
            if match:
                rank = match.group(1).replace(",", "")
                category = match.group(2).strip()
                return f"#{rank} in {category}"

    detail_bullets = tree.xpath('//*[@id="detailBulletsWrapper_feature_div"]//text()')
    if detail_bullets:
        full_text = " ".join(t.strip() for t in detail_bullets)
        match = re.search(r"#([\d,]+)\s+in\s+(.+?)(?:\(|$)", full_text)
        if match:
            rank = match.group(1).replace(",", "")
            category = match.group(2).strip()
            return f"#{rank} in {category}"

    return None


def extract_bsr_numeric(tree: html.HtmlElement) -> Optional[int]:
    bsr = extract_bsr(tree)
    if bsr:
        match = re.search(r"#([\d,]+)", bsr)
        if match:
            return int(match.group(1).replace(",", ""))
    return None


def extract_bsr_category(tree: html.HtmlElement) -> Optional[str]:
    bsr = extract_bsr(tree)
    if bsr:
        match = re.search(r"in\s+(.+)", bsr)
        if match:
            return match.group(1).strip()
    return None


def extract_seller(tree: html.HtmlElement) -> Optional[str]:
    el = tree.xpath('//*[@id="sellerProfileTriggerId"]/text()')
    if el:
        return el[0].strip()
    el = tree.xpath('//*[@id="merchant-info"]//a/text()')
    if el:
        return el[0].strip()
    merchant_info = tree.xpath('//*[@id="merchant-info"]//text()')
    if merchant_info:
        text = " ".join(t.strip() for t in merchant_info if t.strip())
        if "amazon" in text.lower():
            return "Amazon"
    el = tree.xpath('//div[@tabular-attribute-name="Sold by"]//span/text()')
    if el:
        return el[0].strip()
    return None


def extract_fulfillment(tree: html.HtmlElement) -> Optional[str]:
    fulfillment_area = tree.xpath(
        '//*[@id="tabular-buybox"]//text() | '
        '//*[@id="merchant-info"]//text() | '
        '//*[contains(@class,"offer-display-feature-text")]//text()'
    )
    text = " ".join(t.strip().lower() for t in fulfillment_area if t.strip())

    if not text:
        return None

    ships_from_amazon = bool(re.search(r"ships?\s*from[\s\S]{0,30}amazon", text))
    sold_by_amazon = bool(re.search(r"sold\s*by[\s\S]{0,30}amazon", text))

    if ships_from_amazon or sold_by_amazon:
        return "FBA"

    if "easy ship" in text or "easyship" in text:
        return "Easy Ship"

    if re.search(r"ships?\s*from", text) and not ships_from_amazon:
        return "FBM"

    if "fulfilled by amazon" in text or "fulfilment by amazon" in text:
        return "FBA"

    return "FBM"


def extract_deal(tree: html.HtmlElement) -> str:
    """Is the red deal badge on the page? Returns "Yes" / "No".

    **Structure, not vocabulary.** The previous version matched a hardcoded list of
    phrases — 'limited time deal', 'lightning deal', 'prime day deal' — and reported
    "No" for every product during the Freedom Sale, because the badge read
    **"Freedom Sale Deal"** and that string was not on the list. Amazon names each
    sale differently ("Great Indian Festival", "Freedom Sale", whatever is next), so
    a keyword list is guaranteed to go stale and it fails SILENTLY: every row reads
    No, which is indistinguishable from genuinely having no deals.

    Two structural signals are used instead, both taken from the live DOM:

    1. ``#dealBadgeSupportingText`` — the span holding the badge's VISIBLE text. On a
       page with a deal it contains exactly "Freedom Sale Deal"; on one without, the
       span does not exist at all.
    2. ``data-csa-c-painter="dp-deal"`` — Amazon's own marker that the detail page
       painted a deal. Absent when there is no deal.

    Verified against real pages of both kinds:

        with a deal    (B0CY84RYRG): supporting text present, painter present
        without a deal (0143448145): dealBadge div present but BOTH absent

    Note the div itself is always present, which is why its mere existence cannot be
    the test — that is the trap the old countdown-element check fell into.

    **The screen-reader spans are deliberately excluded.** They are `aok-hidden` and
    contain the badge text with UNSUBSTITUTED placeholders —
    "Freedom Sale Deal NO_OF_HOURS hours NO_OF_MINUTES minutes" — because the
    countdown is filled in by JavaScript that never runs here. Reading those would
    work today and would also match a page where the template shipped but no deal was
    active, so the visible span is the honest source.
    """
    # 1. The visible badge text. Most direct: if it is there, the shopper sees a badge.
    visible = " ".join(
        t.strip()
        for t in tree.xpath('//*[@id="dealBadgeSupportingText"]//text()')
        if t.strip()
    )
    # Guard against the placeholder text leaking in through a markup change; a badge
    # whose text is still a template is not a rendered badge.
    if visible and "NO_OF_" not in visible:
        return "Yes"

    # 2. Amazon's own "this page painted a deal" marker, scoped to the badge region so
    #    a deal on a *recommended* product elsewhere on the page cannot trigger it.
    if tree.xpath(
        '//*[@data-feature-name="dealBadge"]//*[@data-csa-c-painter="dp-deal"] | '
        '//span[contains(@class,"dealBadge")][@data-csa-c-painter="dp-deal"]'
    ):
        return "Yes"

    return "No"


def extract_use_by(tree: html.HtmlElement) -> Optional[str]:
    # Primary: Check expiryDate_feature_div (Amazon's dedicated expiry widget)
    expiry_div = tree.xpath('//*[@id="expiryDate_feature_div"]//text()')
    if expiry_div:
        expiry_text = " ".join(t.strip() for t in expiry_div if t.strip())
        # Extract date after "Use by:" or similar label
        match = re.search(
            r"(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{1,2}\s*\w{3,9}\s*\d{2,4})",
            expiry_text, re.IGNORECASE
        )
        if match:
            return match.group(1).strip()
        # If no label prefix, just grab any date-like pattern
        match = re.search(r"(\d{1,2}\s+[A-Z]{3}\s+\d{4})", expiry_text)
        if match:
            return match.group(1).strip()

    # Secondary: Check freshShelfLifeMessage div
    shelf_div = tree.xpath('//*[contains(@id,"freshShelfLife")]//text()')
    if shelf_div:
        shelf_text = " ".join(t.strip() for t in shelf_div if t.strip())
        match = re.search(
            r"(?:use\s*by|best\s*before|expiry)[:\s]*(\d{1,2}[\s/-]\w{3,9}[\s/-]\d{2,4})",
            shelf_text, re.IGNORECASE
        )
        if match:
            return match.group(1).strip()

    # Fallback: Check product detail tables
    detail_texts = tree.xpath(
        '//table[contains(@id,"productDetails")]//text() | '
        '//*[@id="detailBulletsWrapper_feature_div"]//text() | '
        '//*[@id="productDetails_techSpec_section_1"]//text() | '
        '//*[@id="productDetails_detailBullets_sections1"]//text()'
    )
    full_text = " ".join(t.strip() for t in detail_texts if t.strip())

    patterns = [
        r"(?:use\s*by|best\s*before|expiry|expiration|exp\.?\s*date)[:\s]*(\d{1,2}[\s/-]\w{3,9}[\s/-]\d{2,4})",
        r"(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{4}[-/]\d{2}[-/]\d{2})",
        r"(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{1,2}[-/]\d{1,2}[-/]\d{2,4})",
    ]
    for pattern in patterns:
        match = re.search(pattern, full_text, re.IGNORECASE)
        if match:
            return match.group(1).strip()
    return None


def detect_captcha(tree: html.HtmlElement) -> bool:
    captcha_indicators = [
        '//form[@action="/errors/validateCaptcha"]',
        '//img[contains(@src,"captcha")]',
        '//*[contains(text(),"Enter the characters you see below")]',
        '//*[contains(text(),"Type the characters")]',
    ]
    for sel in captcha_indicators:
        if tree.xpath(sel):
            return True
    return False


def detect_dog_page(tree: html.HtmlElement) -> bool:
    dog_indicators = [
        '//*[contains(text(),"looking for was not found")]',
        '//img[contains(@alt,"sorry")]',
        '//*[contains(@class,"a-spacing-base") and contains(text(),"no results")]',
    ]
    for sel in dog_indicators:
        if tree.xpath(sel):
            return True
    return False


def detect_unavailable(tree: html.HtmlElement) -> bool:
    """Detect 'Currently unavailable' / 'out of stock' / suppressed listings."""
    unavail_indicators = [
        '//*[@id="availability"]//text()',
        '//*[@id="availability_feature_div"]//text()',
        '//*[@id="outOfStock"]//text()',
    ]
    for sel in unavail_indicators:
        texts = tree.xpath(sel)
        combined = " ".join(t.strip().lower() for t in texts if t.strip())
        if any(phrase in combined for phrase in (
            "currently unavailable",
            "out of stock",
            "we don't know when or if",
            "unavailable",
        )):
            return True

    # Also check if there's no buybox at all (no "Add to Cart" button and no price)
    has_buybox = bool(
        tree.xpath('//*[@id="add-to-cart-button"]') or
        tree.xpath('//*[@id="buy-now-button"]') or
        tree.xpath('//span[contains(@class,"priceToPay")]')
    )
    # If there's a title but absolutely no buybox, treat as unavailable
    title = tree.xpath('//*[@id="productTitle"]/text()')
    if title and not has_buybox:
        return True

    return False


def parse_product_page(raw_html: str, asin: str) -> dict:
    tree = html.fromstring(raw_html)

    if detect_captcha(tree):
        return {
            "asin": asin,
            "url": f"https://www.amazon.in/dp/{asin}",
            "status": "Blocked (CAPTCHA)",
        }

    if detect_dog_page(tree):
        return {
            "asin": asin,
            "url": f"https://www.amazon.in/dp/{asin}",
            "status": "Not Found",
        }

    title = extract_title(tree)
    if not title:
        return {
            "asin": asin,
            "url": f"https://www.amazon.in/dp/{asin}",
            "status": "Parse Error (no title)",
        }

    # If listing exists but is currently unavailable — keep title/ratings/BSR
    # but explicitly clear price/deal/seller so stale data doesn't persist
    if detect_unavailable(tree):
        return {
            "asin": asin,
            "url": f"https://www.amazon.in/dp/{asin}",
            "title": title,
            "rating": extract_rating(tree),
            "rating_count": extract_rating_count(tree),
            "bsr": extract_bsr(tree),
            "bsr_numeric": extract_bsr_numeric(tree),
            "bsr_category": extract_bsr_category(tree),
            "price": None,       # explicitly clear
            "seller": None,      # explicitly clear
            "fulfillment": None, # explicitly clear
            "deal": "No",        # explicitly clear
            "use_by": extract_use_by(tree),
            "status": "Unavailable",
        }

    return {
        "asin": asin,
        "url": f"https://www.amazon.in/dp/{asin}",
        "title": title,
        "rating": extract_rating(tree),
        "rating_count": extract_rating_count(tree),
        "bsr": extract_bsr(tree),
        "bsr_numeric": extract_bsr_numeric(tree),
        "bsr_category": extract_bsr_category(tree),
        "price": extract_price(tree),
        "seller": extract_seller(tree),
        "fulfillment": extract_fulfillment(tree),
        "deal": extract_deal(tree),
        "use_by": extract_use_by(tree),
        "status": "OK",
    }

File: node-app/server/src/scraper/parsers.ts
/**
 * Product-page extractors, ported from `app/scraper/parsers.py`.
 *
 * **Selector ORDER is the rule, not an implementation detail.** `extractPrice` tries six selectors
 * and takes the first that matches; reordering them changes which price a page reports. The order
 * below is the Python order, verbatim.
 *
 * **Every selection takes `.first()`, and that is load-bearing.** lxml's `tree.xpath(...)` returns
 * a list and the Python code reads `result[0]`; cheerio's `.text()` concatenates the text of
 * *every* match. Measured on the real `B0D817HX57`, whose last-resort price selector matches 21
 * nodes:
 *
 *     $(sel).text()         -> "294.494.597.35294.494.47.597.175.181.450"
 *     $(sel).first().text() -> "294."
 *
 * `₹294..00` is obviously wrong; on a page with two matches the concatenation produces a
 * plausible number instead, which is the dangerous case.
 *
 * XPath `contains(@class,"x")` becomes the CSS attribute-substring selector `[class*="x"]`, and the
 * one union — `//table[contains(@id,"productDetails")]//tr | //div[@id="…"]//li` — becomes a comma
 * list, verified to return the same 11 rows on a real page.
 */
import * as cheerio from "cheerio";
import type { CheerioAPI } from "cheerio";

import {
  detectCaptcha,
  detectDogPage,
  detectUnavailable,
  hasTitle,
} from "./guards.js";

/** What one scraped page yields. Mirrors the Python dict, with the statuses as a closed union. */
export type ScrapeStatus =
  | "OK"
  | "Blocked (CAPTCHA)"
  | "Not Found"
  | "Parse Error (no title)"
  | "Unavailable";

export interface ProductRow {
  asin: string;
  url: string;
  status: ScrapeStatus;
  title: string | null;
  price: string | null;
  rating: string | null;
  ratingCount: string | null;
  bsr: string | null;
  bsrNumeric: number | null;
  bsrCategory: string | null;
  seller: string | null;
  fulfillment: string | null;
  deal: string;
  useBy: string | null;
}

/** The six price selectors, in the order the Python file tries them. Order IS the rule. */
const PRICE_SELECTORS = [
  'span[class*="priceToPay"] span[class*="a-price-whole"]',
  '#corePrice_feature_div span[class*="a-price-whole"]',
  "span#priceblock_ourprice",
  "span#priceblock_dealprice",
  '#apex_offerDisplay_desktop span[class*="a-price-whole"]',
  'span[class*="a-price"] span[class*="a-price-whole"]',
] as const;

export function extractTitle($: CheerioAPI): string | null {
  const text = $("#productTitle").first().text().trim();
  return text || null;
}

/**
 * The price as `₹123` or `₹123.45`, or null.
 *
 * The rupee symbol and the whole/fraction split are Amazon's markup: the integer part and the
 * decimals live in separate spans, so the fraction is looked up with the SAME selector that
 * matched the whole — swapping the class name, exactly as the Python does.
 */
export function extractPrice($: CheerioAPI): string | null {
  for (const selector of PRICE_SELECTORS) {
    const whole = $(selector).first().text().trim().replace(/,/g, "");
    if (!whole) continue;
    // Some of these selectors match a full price string rather than a whole-part span
    // (`#priceblock_ourprice`), in which case it already carries the symbol.
    if (whole.includes("₹")) return whole;
    const fraction = $(selector.replace("a-price-whole", "a-price-fraction"))
      .first()
      .text()
      .trim();
    // The whole-part span's text often ends in the decimal separator ("294."), so it is stripped
    // before the fraction is appended — otherwise the result reads "₹294..00".
    const cleaned = whole.replace(/[.,]$/, "");
    return fraction ? `₹${cleaned}.${fraction}` : `₹${cleaned}`;
  }
  return null;
}

export function extractRating($: CheerioAPI): string | null {
  const fromTitle = $("#acrPopover").first().attr("title") ?? "";
  const m1 = /([\d.]+)/.exec(fromTitle);
  if (m1?.[1]) return m1[1];

  const outOf = $('span[data-hook="rating-out-of-text"]').first().text();
  const m2 = /([\d.]+)/.exec(outOf);
  if (m2?.[1]) return m2[1];

  const iconText = $('i[class*="a-icon-star"] span').first().text();
  const m3 = /([\d.]+)/.exec(iconText);
  return m3?.[1] ?? null;
}

export function extractRatingCount($: CheerioAPI): string | null {
  const selectors = [
    "#acrCustomerReviewText",
    "#acrCustomerReviewLink span",
    'span[data-hook="total-review-count"]',
  ];
  for (const selector of selectors) {
    // Each MATCH is tested, not just the first, because the Python version loops the node list —
    // `#acrCustomerReviewLink span` matches several spans and only some carry the number.
    const nodes = $(selector).toArray();
    for (const node of nodes) {
      const match = /([\d,]+)/.exec($(node).text());
      if (match?.[1]) return match[1].replace(/,/g, "");
    }
  }
  return null;
}

/** `#1,234 in Category` from any of the three places Amazon puts it. */
export function extractBsr($: CheerioAPI): string | null {
  const pattern = /#([\d,]+)\s+in\s+(.+?)(?:\(|$)/;

  const salesRank = $("#SalesRank").text().replace(/\s+/g, " ").trim();
  if (salesRank) {
    const match = pattern.exec(salesRank);
    if (match?.[1] && match[2]) {
      return `#${match[1].replace(/,/g, "")} in ${match[2].trim()}`;
    }
  }

  // The XPath union, as a CSS comma list. Verified to return the same 11 rows on a real page.
  const rows = $(
    'table[id*="productDetails"] tr, #detailBulletsWrapper_feature_div li',
  ).toArray();
  for (const row of rows) {
    const text = $(row).text().replace(/\s+/g, " ").trim();
    if (
      text.includes("Best Sellers Rank") ||
      text.toLowerCase().includes("best seller")
    ) {
      const match = pattern.exec(text);
      if (match?.[1] && match[2]) {
        return `#${match[1].replace(/,/g, "")} in ${match[2].trim()}`;
      }
    }
  }

  const bullets = $("#detailBulletsWrapper_feature_div")
    .text()
    .replace(/\s+/g, " ")
    .trim();
  if (bullets) {
    const match = pattern.exec(bullets);
    if (match?.[1] && match[2]) {
      return `#${match[1].replace(/,/g, "")} in ${match[2].trim()}`;
    }
  }

  return null;
}

export function extractBsrNumeric($: CheerioAPI): number | null {
  const bsr = extractBsr($);
  if (!bsr) return null;
  const match = /#([\d,]+)/.exec(bsr);
  if (!match?.[1]) return null;
  return Number.parseInt(match[1].replace(/,/g, ""), 10);
}

export function extractBsrCategory($: CheerioAPI): string | null {
  const bsr = extractBsr($);
  if (!bsr) return null;
  const match = /in\s+(.+)/.exec(bsr);
  return match?.[1]?.trim() ?? null;
}

export function extractSeller($: CheerioAPI): string | null {
  const trigger = $("#sellerProfileTriggerId").first().text().trim();
  if (trigger) return trigger;

  const merchantLink = $("#merchant-info a").first().text().trim();
  if (merchantLink) return merchantLink;

  const merchantInfo = $("#merchant-info").text().trim();
  if (merchantInfo.toLowerCase().includes("amazon")) return "Amazon";

  const tabular = $('div[tabular-attribute-name="Sold by"] span')
    .first()
    .text()
    .trim();
  return tabular || null;
}

/**
 * `FBA` / `FBM` / `Easy Ship`, or null when the buy box says nothing.
 *
 * The rule order is the rule: "ships from Amazon" or "sold by Amazon" wins before the Easy Ship
 * check, and a bare "ships from" that is NOT Amazon means FBM. The `[\s\S]{0,30}` gap in the
 * Python regexes matters — the words are separated by markup, so the two are not adjacent.
 */
export function extractFulfillment($: CheerioAPI): string | null {
  const text = $(
    '#tabular-buybox, #merchant-info, [class*="offer-display-feature-text"]',
  )
    .text()
    .replace(/\s+/g, " ")
    .toLowerCase()
    .trim();

  if (!text) return null;

  const shipsFromAmazon = /ships?\s*from[\s\S]{0,30}amazon/.test(text);
  const soldByAmazon = /sold\s*by[\s\S]{0,30}amazon/.test(text);
  if (shipsFromAmazon || soldByAmazon) return "FBA";

  if (text.includes("easy ship") || text.includes("easyship")) return "Easy Ship";
  if (/ships?\s*from/.test(text) && !shipsFromAmazon) return "FBM";
  if (
    text.includes("fulfilled by amazon") ||
    text.includes("fulfilment by amazon")
  ) {
    return "FBA";
  }
  return "FBM";
}

/**
 * Is the red deal badge on the page? `"Yes"` / `"No"`.
 *
 * **Structure, not vocabulary.** An earlier Python version matched a hardcoded phrase list and
 * reported "No" for every product during the Freedom Sale, because the badge read "Freedom Sale
 * Deal" and that string was not on the list. Amazon renames the sale every few months and the
 * failure is SILENT — every row reads No, which is indistinguishable from having no deals.
 *
 * Three traps, all of them live:
 *
 * 1. `#dealBadge_feature_div` is on EVERY product page, deal or not, so its presence cannot be
 *    the test.
 * 2. The `aok-hidden` screen-reader spans hold the badge text with the countdown
 *    unsubstituted — `"Freedom Sale Deal NO_OF_HOURS hours"` — because that JavaScript never runs
 *    here. Text containing `NO_OF_` is rejected. Same trap as the FBA transportation quote coming
 *    back as the literal string `"$cost.amount"`.
 * 3. Ad carousels carry deal markup for OTHER products — 7 such mentions on one measured page — so
 *    the painter check is scoped to the badge region.
 */
export function extractDeal($: CheerioAPI): string {
  const visible = $("#dealBadgeSupportingText").text().replace(/\s+/g, " ").trim();
  if (visible && !visible.includes("NO_OF_")) return "Yes";

  const painter = $(
    '[data-feature-name="dealBadge"] [data-csa-c-painter="dp-deal"], ' +
      'span[class*="dealBadge"][data-csa-c-painter="dp-deal"]',
  );
  if (painter.length > 0) return "Yes";

  return "No";
}

/** The use-by / best-before date as Amazon prints it, or null. */
export function extractUseBy($: CheerioAPI): string | null {
  const expiry = $("#expiryDate_feature_div").text().replace(/\s+/g, " ").trim();
  if (expiry) {
    const labelled =
      /(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{1,2}\s*\w{3,9}\s*\d{2,4})/i.exec(
        expiry,
      );
    if (labelled?.[1]) return labelled[1].trim();
    const bare = /(\d{1,2}\s+[A-Z]{3}\s+\d{4})/.exec(expiry);
    if (bare?.[1]) return bare[1].trim();
  }

  const shelf = $('[id*="freshShelfLife"]').text().replace(/\s+/g, " ").trim();
  if (shelf) {
    const match =
      /(?:use\s*by|best\s*before|expiry)[:\s]*(\d{1,2}[\s/-]\w{3,9}[\s/-]\d{2,4})/i.exec(
        shelf,
      );
    if (match?.[1]) return match[1].trim();
  }

  const details = $(
    'table[id*="productDetails"], #detailBulletsWrapper_feature_div, ' +
      "#productDetails_techSpec_section_1, #productDetails_detailBullets_sections1",
  )
    .text()
    .replace(/\s+/g, " ")
    .trim();

  const patterns = [
    /(?:use\s*by|best\s*before|expiry|expiration|exp\.?\s*date)[:\s]*(\d{1,2}[\s/-]\w{3,9}[\s/-]\d{2,4})/i,
    /(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{4}[-/]\d{2}[-/]\d{2})/i,
    /(?:use\s*by|best\s*before|expiry|expiration)[:\s]*(\d{1,2}[-/]\d{1,2}[-/]\d{2,4})/i,
  ];
  for (const pattern of patterns) {
    const match = pattern.exec(details);
    if (match?.[1]) return match[1].trim();
  }
  return null;
}

function emptyRow(asin: string, status: ScrapeStatus): ProductRow {
  return {
    asin,
    url: `https://www.amazon.in/dp/${asin}`,
    status,
    title: null,
    price: null,
    rating: null,
    ratingCount: null,
    bsr: null,
    bsrNumeric: null,
    bsrCategory: null,
    seller: null,
    fulfillment: null,
    deal: "No",
    useBy: null,
  };
}

/**
 * One page to one row. **The guards run first and they decide what the row may contain.**
 *
 * The unavailable branch keeps title, rating and BSR while explicitly nulling price, seller,
 * fulfillment and deal — so a listing that cannot be bought cannot carry a price. Measured: on
 * `B0D817HX57` the price chain would otherwise return a real ₹294 belonging to a *different*
 * product in an ad carousel, and the row would look completely normal.
 */
export function parseProductPage(rawHtml: string, asin: string): ProductRow {
  const $ = cheerio.load(rawHtml);

  if (detectCaptcha($)) return emptyRow(asin, "Blocked (CAPTCHA)");
  if (detectDogPage($)) return emptyRow(asin, "Not Found");
  if (!hasTitle($)) return emptyRow(asin, "Parse Error (no title)");

  const title = extractTitle($);

  if (detectUnavailable($)) {
    return {
      ...emptyRow(asin, "Unavailable"),
      title,
      rating: extractRating($),
      ratingCount: extractRatingCount($),
      bsr: extractBsr($),
      bsrNumeric: extractBsrNumeric($),
      bsrCategory: extractBsrCategory($),
      useBy: extractUseBy($),
    };
  }

  return {
    asin,
    url: `https://www.amazon.in/dp/${asin}`,
    status: "OK",
    title,
    price: extractPrice($),
    rating: extractRating($),
    ratingCount: extractRatingCount($),
    bsr: extractBsr($),
    bsrNumeric: extractBsrNumeric($),
    bsrCategory: extractBsrCategory($),
    seller: extractSeller($),
    fulfillment: extractFulfillment($),
    deal: extractDeal($),
    useBy: extractUseBy($),
  };
}
